File: 2024/18/18.go
package main

import (
	"bufio"
	"image"
	"os"
	"slices"
	"strconv"
	"strings"
)

var XMAX, YMAX = 70, 70
var DIRECTIONS = []image.Point{{0, -1}, {1, 0}, {0, 1}, {-1, 0}}

type Visit struct {
	pos  image.Point
	path []image.Point
}

func atoi(s string) int {
	i, _ := strconv.Atoi(s)
	return i
}

func parseInput() []image.Point {
	positions := []image.Point{}
	scanner := bufio.NewScanner(os.Stdin)

	for scanner.Scan() {
		pair := strings.Split(scanner.Text(), ",")
		positions = append(positions, image.Point{atoi(pair[0]), atoi(pair[1])})
	}
	return positions
}

func inBounds(pos image.Point) bool {
	return pos.X >= 0 && pos.X <= XMAX && pos.Y >= 0 && pos.Y <= YMAX
}

func findPath(falling_bytes []image.Point) []image.Point {
	start, end := image.Point{0, 0}, image.Point{XMAX, YMAX}
	positions := []Visit{{start, []image.Point{}}}
	visited := map[image.Point]bool{}
	fallen := map[image.Point]bool{}
	for _, f := range falling_bytes {
		fallen[f] = true
	}

	var visit Visit

	for len(positions) > 0 {
		visit, positions = positions[0], positions[1:]

		if visit.pos == end {
			return append(visit.path, visit.pos)
		}
		if visited[visit.pos] {
			continue
		}
		visited[visit.pos] = true

		for _, d := range DIRECTIONS {
			np := visit.pos.Add(d)
			if inBounds(np) && !fallen[np] {
				positions = append(positions, Visit{np, append(slices.Clone(visit.path), visit.pos)})
			}
		}
	}
	return []image.Point{}
}

func searchFirstBrokenPath(positions []image.Point) image.Point {
	min, max := 1024, len(positions)-1
	for min < max-1 {
		center := (min + max) / 2
		path := findPath(positions[0 : center+1])
		if len(path) > 0 {
			min = center + 1
		} else {
			max = center - 1
		}
	}
	return positions[max]
}

func main() {
	positions := parseInput()

	path := findPath(positions[0:1024])
	println(len(path) - 1)

	brokenPathByte := searchFirstBrokenPath(positions)
	println(strconv.Itoa(brokenPathByte.X) + "," + strconv.Itoa(brokenPathByte.Y))
}

File: 2024/18/18.rb
#!/usr/bin/env ruby

require 'matrix'
V = Vector

DIRECTIONS = [V[0, -1], V[1, 0], V[0, 1], V[-1, 0]].freeze
X, Y = [(0..70), (0..70)]

def find_path(falling_bytes)
  start, ending = [V[X.min,Y.min], V[X.max,Y.max]]
  positions = [[start, []]]
  visited = Set.new
  fallen = falling_bytes.to_set

  while positions.any?
    position, path = positions.shift

    return path + [position] if position == ending
    next if visited.include? position
    visited << position

    DIRECTIONS.each do |d|
      np = position + d
      next unless X.include? np[0]
      next unless Y.include? np[1]
      positions << [np, path + [position]] unless fallen.include? np
    end
  end
  []
end

positions = $stdin.readlines.map { V[*_1.split(',').map(&:to_i)] }

path = find_path(positions[0...1024])
puts path.size - 1

broken_path_byte = (0...positions.size).bsearch do |i|
  find_path(positions[0..i]).size == 0
end
puts positions[broken_path_byte].to_a.map(&:to_s).join(',')
